### ISSUE TITLE
NoReturn on methods appears to be ignored.

### ISSUE THREAD

**Issue by rphes** [microsoft/pyright#2206]
**Describe the bug**
Declaring a method with a return type of `NoReturn` and then calling that method on an instance will cause pyright to report false positives on the calling code.

I've seen a couple of similar reports on the issue tracker, but nothing exactly like this.

**To Reproduce**
Consider the following snippet:
```python
from typing import cast, NoReturn, Callable

class Foo:
    def raise_error(self) -> NoReturn:
        raise ValueError

foo = Foo()

def bar(x: bool) -> str:
    if x:
        return "true"

    foo.raise_error()

def bar2(x: bool) -> str:
    if x:
        return "true"

    cast(Callable[[], NoReturn], foo.raise_error)()
```

Both `bar` and `bar2` report:

```
Function with declared type of "str" must return value
Type "None" cannot be assigned to type "str"
```

Interestingly, calling `reveal_type` on either function shows the expected type.

**Expected behavior**
Both functions should successfully typecheck.

**VS Code extension or command-line**
I'm running pyright 1.1.162 in neovim using coc-nvim and its coc-pyright plugin.

**Comment by rphes** [comment]
Thanks for the explanation. I did come across #1562, but it wasn't obvious to me then that it describes the same issue. Out of curiosity: why does casting the method as I did in `bar2` not help? Is `cast` ignored while building the code flow graph, whereas type annotations are not?

**Comment by erictraut** [comment]
The code that evaluates the reachability within a code flow graph does a form of lightweight type evaluation that doesn't rely on any code flow analysis (hence, why it cannot infer types). This lightweight type evaluation works only for expression patterns that are found commonly in these circumstances. The expression `cast(Callable[[], NoReturn], foo.raise_error)` is way too complex for this lightweight type evaluation logic.

**Comment by erictraut** [comment]
I found a way to improve the logic. Since `foo` is a variable in a different scope, it's safe to infer its type when walking the code flow graph of the inner scope. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.165, which I just published. It will also be included in the next release of pylance.
